src/QsCompiler/SyntaxProcessor/CapabilityInference/TypeAnalyzer.fs

```diff
@@ -0,0 +1,210 @@
+// Licensed under the MIT License.
+
+module Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference.TypeAnalyzer
+
+open Microsoft.Quantum.QsCompiler
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.Diagnostics
+open Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference
+open Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference.ContextRef
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+open Microsoft.Quantum.QsCompiler.Transformations.QsCodeOutput
+
+type Construct =
+    | Literal
+    | Mutable
+    | Param
+    | Return
+    | Conditional
+
+type Scenario =
+    | UseBigInt
+    | StringNotArgumentToMessage
+    | ConditionalOrMutable of ResolvedType
+    | EntryPointParam of ResolvedType
+    | EntryPointReturn of ResolvedType
+
+type Context = { IsEntryPoint: bool; StringLiteralsOk: bool }
+
+let shallowFindScenario context construct (ty: ResolvedType) =
+    match construct with
+    | Literal ->
+        match ty.Resolution with
+        | BigInt -> Some UseBigInt
+        | String when not context.StringLiteralsOk -> Some StringNotArgumentToMessage
+        | _ -> None
+    | Conditional
+    | Mutable -> ConditionalOrMutable ty |> Some
+    | Param when context.IsEntryPoint -> EntryPointParam ty |> Some
+    | Param -> None
+    | Return when context.IsEntryPoint -> EntryPointReturn ty |> Some
+    | Return -> None
+
+let findScenarios context construct (ty: ResolvedType) =
+    ty.ExtractAll(shallowFindScenario context construct >> Option.toList >> Seq.ofList)
+
+let scenarioCapability scenario =
+    let classical =
+        match scenario with
+        | UseBigInt -> ClassicalCapability.full
+        | StringNotArgumentToMessage -> ClassicalCapability.full
+        | ConditionalOrMutable ty ->
+            match ty.Resolution with
+            | Bool
+            | Int -> ClassicalCapability.integral
+            | _ -> ClassicalCapability.full
+        | EntryPointParam ty ->
+            match ty.Resolution with
+            | TupleType _ -> ClassicalCapability.empty
+            | Bool
+            | Int -> ClassicalCapability.integral
+            | _ -> ClassicalCapability.full
+        | EntryPointReturn ty ->
+            match ty.Resolution with
+            | UnitType
+            | Result
+            | TupleType _
+            | ArrayType _ -> ClassicalCapability.empty
+            | Bool
+            | Int -> ClassicalCapability.integral
+            | _ -> ClassicalCapability.full
+
+    TargetCapability.withClassical classical TargetCapability.bottom
+
+let shallowTypeName (ty: ResolvedType) =
+    match ty.Resolution with
+    | ArrayType _ -> "array"
+    | TupleType _ -> "tuple"
+    | QsTypeKind.Operation _ -> "operation"
+    | QsTypeKind.Function _ -> "function"
+    | _ -> SyntaxTreeToQsharp.Default.ToCode ty
+
+let describeScenario =
```

Comment:
This seems like the error string is not split between the diagnostics file and here. I'd prefer to absorb it all into the diagnostics file.

Reply:
Yes, but I didn't want to create a ton of different error codes for each scenario. Since the scenarios we are looking for are very fluid and will likely change over time, it seemed better to have one code for an unsupported classical capability, with the error message describing the specific problem.

I would prefer not to use strings here, too, but that's difficult to do unless microsoft/qsharp-compiler#1025 is implemented.

Reply:
Alright, I see your point about an expected number of changes in the next couple of weeks. Let's defer that then to be processed with 1025. Could you please add a comment though to the strings in each of the analyzers pointing that out and linking issue 1025?